src/main/kotlin/org/axonframework/intellij/ide/plugin/resolving/MessageCreationResolver.kt

```diff
@@ -0,0 +1,94 @@
+ * It does this by searching for constructor references of compatible payloads. Inheritance is supported.
+ *
+ * Results are cached based on the Psi modifications of IntelliJ. This means the calculations are invalidated when
+ * the PSI is modified (code is edited) or is collected by the garbage collector.
+ */
+class MessageCreationResolver(private val project: Project) {
+    private val psiFacade = JavaPsiFacade.getInstance(project)
+    private val handlerResolver = project.getService(MessageHandlerResolver::class.java)
+    private val constructorsByPayloadCache = ConcurrentHashMap<String, CachedValue<List<MessageCreator>>>()
+
+
+    fun getCreatorsForPayload(payloadQualifiedClassName: String): List<MessageCreator> {
+        val cache = constructorsByPayloadCache.getOrPut(payloadQualifiedClassName) {
+            project.createCachedValue {
+                resolveCreatorForFqn(payloadQualifiedClassName)
+            }
+        }
+        return cache.value
+    }
+
+    /**
+     * Finds the already constructed/found creator in the caches. Useful for quick filtering in line marker popups.
+     */
+    fun findCreatorByElement(element: PsiElement): MessageCreator? {
+        return constructorsByPayloadCache.values.filter { it.hasUpToDateValue() }
+                .flatMap { it.value }
+                .firstOrNull { it.element == element }
+    }
+
+    private fun resolveCreatorForFqn(qualifiedName: String): List<MessageCreator> {
+        val classesForQualifiedName = handlerResolver.findAllHandlers()
+                .map { it.payloadFullyQualifiedName }.distinct()
+                .filter { areAssignable(project, qualifiedName, it) }
+        return resolveCreatorsForFqns(classesForQualifiedName)
+    }
+
+    /**
+     * This action is VERY expensive. Should only be used if the user does not depend on it or is expected to wait.
+     * For example, when creating an Event Modeling board based on this info.
+     */
+    fun resolveAllCreators(): List<MessageCreator> {
+        val handlers = handlerResolver.findAllHandlers()
+        val handlerTypes = handlers.map { it.payloadFullyQualifiedName }.distinct()
+
+        return resolveCreatorsForFqns(handlerTypes)
+    }
+
+    private fun resolveCreatorsForFqns(fqns: List<String>): List<MessageCreator> {
+        return fqns.flatMap { typeFqn ->
+            val clazz = psiFacade.findClass(typeFqn, project.axonScope()) ?: return@flatMap emptyList()
+            clazz.constructors
+                    .flatMap { MethodReferencesSearch.search(it, project.axonScope(), true) }
+                    .map { ref -> createCreator(ref.element, typeFqn) }
+                    .distinct()
+        }
+    }
+
+    private fun createCreator(element: PsiElement, qualifiedName: String): MessageCreator {
+        return DefaultMessageCreator(element, qualifiedName, findParentHandler(element))
+    }
+
+    private fun findParentHandler(element: PsiElement): Handler? {
```

Comment:
Limitation; if the creator is in a nested function, it won't show up here.
Might need to do a recursive references search. But this might be very expensive.

Might be good for a new issue/PR to try out experimentally

Reply:
Other idea; this can be calculated lazily when clicking a line marker as well

Reply:
Lazy evaluation does not work; the line marker lazy value is not invalidated on psi changes, so you will never see new items. 
Implemented recursive parent handler finding